eth/watchers/senderwatcher_test.go

```diff
@@ -0,0 +1,469 @@
+	"github.com/livepeer/go-livepeer/eth"
+	"github.com/livepeer/go-livepeer/eth/blockwatch"
+	"github.com/livepeer/go-livepeer/pm"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestGetAndSetSenderInfo(t *testing.T) {
+	assert := assert.New(t)
+	sw := &SenderWatcher{
+		senders: make(map[ethcommon.Address]*pm.SenderInfo),
+		lpEth: &eth.StubClient{
+			SenderInfo: &pm.SenderInfo{
+				Deposit: big.NewInt(10),
+				Reserve: big.NewInt(5),
+			},
+		},
+	}
+
+	info := &pm.SenderInfo{
+		Deposit:      big.NewInt(1000000000000000000),
+		Reserve:      big.NewInt(1000000000000000000),
+		ReserveState: pm.NotFrozen,
+	}
+	sender := pm.RandAddress()
+
+	sw.setSenderInfo(sender, info)
+	entry, err := sw.GetSenderInfo(sender)
+	assert.Nil(err)
+	assert.Equal(sw.senders[sender], info)
+	assert.Equal(info, entry)
+
+	// If map entry is nil and we call get, make RPC call and set the returned data
+	sw.Clear(sender)
+	info, err = sw.GetSenderInfo(sender)
+	assert.Nil(err)
+	assert.Zero(info.Deposit.Cmp(big.NewInt(10)))
+	assert.Zero(info.Reserve.Cmp(big.NewInt(5)))
+	info = sw.senders[sender]
+	assert.Zero(info.Deposit.Cmp(big.NewInt(10)))
+	assert.Zero(info.Reserve.Cmp(big.NewInt(5)))
+}
+
+func TestWatchAndStop(t *testing.T) {
+	assert := assert.New(t)
+	lpEth := &eth.StubClient{}
+	watcher := &stubBlockWatcher{}
+	sw, err := NewSenderWatcher(stubTicketBrokerAddr, watcher, lpEth)
+	assert.Nil(err)
+
+	go sw.Watch()
+	time.Sleep(2 * time.Millisecond)
+	sw.Stop()
+	time.Sleep(2 * time.Millisecond)
+	assert.True(watcher.sub.unsubscribed)
+}
+
+func TestSenderWatcher_HandleLog(t *testing.T) {
+	lpEth := &eth.StubClient{}
+	watcher := &stubBlockWatcher{}
+	rw, err := NewSenderWatcher(stubTicketBrokerAddr, watcher, lpEth)
+	require.Nil(t, err)
+
+	assert := assert.New(t)
+
+	// Test unknown event
+	log := newStubBaseLog()
+	log.Topics = []ethcommon.Hash{ethcommon.BytesToHash([]byte("foo"))}
+
+	err = rw.handleLog(log)
+	assert.Nil(err)
+}
+
+func TestFundDepositEvent(t *testing.T) {
+	assert := assert.New(t)
+	senderInfo := &pm.SenderInfo{
+		Deposit: big.NewInt(10),
+		Reserve: big.NewInt(5),
+	}
+	lpEth := &eth.StubClient{
+		SenderInfo: senderInfo,
+	}
+	watcher := &stubBlockWatcher{}
+	sw, err := NewSenderWatcher(stubTicketBrokerAddr, watcher, lpEth)
+	assert.Nil(err)
+
+	// We only keep cache for addresses that exist in the map
+	_, err = sw.GetSenderInfo(stubSender)
+	assert.Nil(err)
+
+	header := defaultMiniHeader()
+	newDepositEvent := newStubDepositFundedLog()
+	header.Logs = append(header.Logs, newDepositEvent)
+
+	blockEvent := &blockwatch.Event{
+		Type:        blockwatch.Added,
+		BlockHeader: header,
+	}
+
+	go sw.Watch()
+	defer sw.Stop()
+	time.Sleep(2 * time.Millisecond)
+
+	// If map entry is nil , make RPC call and init the map entry
+	watcher.sink <- []*blockwatch.Event{blockEvent}
+	time.Sleep(2 * time.Millisecond)
+	info := sw.senders[stubSender]
+	assert.Equal(info, senderInfo)
```

Comment:
I'm not sure what you mean here, could you please elaborate?

 In the cases we parse the event logs we are making checks against the event log data itself. 
When we make an RPC call we check against SenderInfo, yes the cached pointer will be updated with the stubbed out RPC call values, but isn't that the goal here?

Reply:
My original comment was a little off-base - let me rephrase. The test was not technically incorrect - it checked that the values referenced by a cached pointer changed whenever the values referenced by the pointer changed. But, I found reading the test a bit confusing.

Consider this code snippet when f06a0ec was the tip of the branch:

```
// We only keep cache for addresses that exist in the map
_, err = sw.GetSenderInfo(stubSender)
assert.Nil(err)

// Code omitted

// If map entry is nil , make RPC call and init the map entry
watcher.sink <- []*blockwatch.Event{blockEvent}
time.Sleep(2 * time.Millisecond)
info := sw.senders[stubSender]
assert.Equal(info, senderInfo)
```

- `sw.GetSenderInfo()` will set the value for `stubSender` in the cache to `senderInfo`
- `watcher.sink <- []*blockwatch.Event{blockEvent}` will send a `DepositFunded` event to be processed by the SenderWatcher. Since, the address in the event is `stubSender`, the SenderWatcher should increase the `Deposit` field for `stubSender` in the cache
- My original reaction when reading the `assert.Equal(info, senderInfo)` was: "why isn't this failing if `info.Deposit` should be increased?". The answer to my question was that `info` and `senderInfo` reference the same pointer so they will *always* be equal in an assertion. `info.Deposit` is being increased here, but IMO the assertion makes it look like we're checking that `info.Deposit` did not increase

I just wanted to point this out because I think it can lead to confusing test semantics. See [this comment](https://github.com/livepeer/go-livepeer/pull/1054/files#r320357953) for an example where it might lead to confusion.